ocaml/xapi/xapi_pgpu.ml

```diff
@@ -48,9 +48,17 @@ let maybe_fetch_compatibility_metadata ~__context ~pgpu_pci =
     []
 
 let populate_compatibility_metadata ~__context ~pgpu ~pgpu_pci =
-  let () = Db.PGPU.set_compatibility_metadata ~__context ~self:pgpu
-    ~value:(maybe_fetch_compatibility_metadata ~__context ~pgpu_pci)
-  in ()
+  let this = "populate_compatibility_metadata" in
+  try
+    let value = fetch_compatibility_metadata ~__context ~pgpu_pci in
+    Db.PGPU.set_compatibility_metadata ~__context ~self:pgpu ~value
+  with
+  | Gpumon_interface.NvmlInterfaceNotAvailable ->
```

Comment:
What will happen if you replace an nvidia pGPU, which has `compatibility_metadata`, which a pGPU of another brand, in the same PCI slot?

Reply:
If you replace a card, you need a reboot - and this means, you should be able to get the correct data because no VM is using it. I believe the risk that we have stale data is very small and a much better trade-off than not having compatibility data when we need it.

Reply:
If the new card is a non-nvidia one, then gpumon will raise `Gpumon_interface.NvmlInterfaceNotAvailable ` and still we want the metadata to be cleared.

Reply:
It may be that the entire PGPU object is recreated in this case, but it would be good to check.

Reply:
I checked in the code: it looks like you indeed get new PCI and PGPU objects if you replace the card with a different kind of card.

Reply:
I have updated the commit message to explain why I think this is save.